File: LeetCode/84.java
import java.util.Stack;

class Solution
{
    public int largestRectangleArea(int[] heights) {
        int result = 0;
        for (int i = 0, len = heights.length; i < len; i++) {
            for (int j = i, min = Integer.MAX_VALUE; j < len; j++) {
                min = Math.min(min, heights[j]);
                result = Math.max(result, min * (j - i + 1));
            }
        }
        return result;
    }

    // 单调递增栈
    public int largestRectangleArea1(int[] heights) {
        int[] list = new int[heights.length + 1];
        System.arraycopy(heights, 0, list, 0, heights.length);
        list[heights.length] = 0;
        Stack<Integer> stack = new Stack<>();
        int result = 0;
        for (int i = 0, len = list.length; i < len; i++) {
            while (!stack.isEmpty() && list[i] < list[stack.peek()]) {
                int curIndex = stack.pop();
                int acreage = list[curIndex] * (!stack.isEmpty() ? (i - stack.peek() - 1) : i);
                result = Math.max(result, acreage);
            }
            stack.push(i);
        }
        return result;
    }

    public int largestRectangleArea2(int[] heights) {
        if (heights == null || heights.length == 0) return 0;
        int n = heights.length;
        int[] left = new int[n];
        int[] right = new int[n];
        left[0] = -1;
        right[n - 1] = n;
        int result = 0, temp;
        // 左侧比当前柱子矮的第一根柱子的下标
        for (int i = 1; i < n; i++) {
            temp = i - 1;
            while (temp >= 0 && heights[temp] >= heights[i]) temp = left[temp];
            left[i] = temp;
        }
        // 右侧比当前柱子矮的第一根柱子的下标
        for (int i = n - 2; i >= 0; i--) {
            temp = i + 1;
            while (temp < n && heights[temp] >= heights[i]) temp = right[temp];
            right[i] = temp;
        }
        // right[i] - left[i] - 1 这个区间内所有柱子都比当前柱子高
        for (int i = 0; i < n; i++)
            result = Math.max(result, (right[i] - left[i] - 1) * heights[i]);
        return result;
    }
}

File: LeetCode/84.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution
{
public:
    //这个题更最大的盛水面积有点像，必须思考的是最大的面积究竟是在什么集合当中
    //1.显然，最大面积一定产生于以当前柱子为最大高度所能覆盖的面积集合中
    //2.其次要思考面积的计算，即等于当前柱子的高度*覆盖的宽度，覆盖的宽度等于其左右第一个小于其高度的柱子的下标之差
    //3.实现方式，因为需要考虑的只有三个柱子，即当前，左边比当前高度小的第一个柱子，右边比当前高度小的第一个柱子
    //因此可以用单调栈的数据结构来实现，即大于栈顶的高度入栈，小于栈顶高度出栈，并计算当前高度的面积，此时三个
    //柱子分别是栈顶高度，紧邻栈顶的高度，以及遍历的当前高度
    //时间复杂度O(n),空间复杂度O(n)
    int largestRectangleArea(vector<int> &heights)
    {
        heights.push_back(0); //保证右边界为最后一个柱子时，方便计算
        stack<int> S;
        int maxarea(0);
        for (int i = 0; i < heights.size(); i++)
        {
            while (!S.empty() && heights[i] < heights[S.top()])
            {
                int curheight = S.top();
                S.pop();
                //判断一下如果栈顶柱子为空，则令栈顶元素为0
                int area = heights[curheight] * (!S.empty() ? (i - S.top() - 1) : i);
                maxarea = max(maxarea, area);
            }
            S.push(i);
        }
        return maxarea;
    }
};